src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class C
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { IsOk: true } c:
+                return c.Value;
+            default:
+                throw null!;
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_WithTopLevelDeclaration_WithType()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class C
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case C { IsOk: true } c: // note: has type
+                return c.Value;
+            case C c:
+                return c.Value; // 1
+            default:
+                throw null!;
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return c.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "c.Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_BinaryPatternTest()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class C
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public bool IsRandom { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        if (this is { IsOk: true } and { IsRandom: true })
+        {
+            return Value;
+        }
+
+        if (this is { IsOk: true } or { IsRandom: true })
+        {
+            return Value; // 1
+        }
+
+        return Value; // 2
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (21,20): warning CS8603: Possible null reference return.
+                //             return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(21, 20),
+                // (24,16): warning CS8603: Possible null reference return.
+                //         return Value; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(24, 16)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_BinaryPatternTest_SideEffectOnStruct()
```

Comment:
>MemberNotNullWhenTrue_BinaryPatternTest_SideEffectOnStruct [](start = 20, length = 58)

Is this test needed? (Does pattern matching differentiate members that have side-effects?) And where is the side-effect?

Reply:
Same comment as `MemberNotNullWhenTrue_RecursivePattern_SideEffectOnStruct`, except that we're dealing with an `is` rather than a `switch`.

---
In reply to: [573113257](https://github.com/dotnet/roslyn/pull/51001#discussion_r573113257) [](ancestors = 573113257)